lib/AST/ASTContext.cpp

```diff
@@ -8500,6 +8527,8 @@ QualType ASTContext::mergeTypes(QualType LHS, QualType RHS,
            "Equivalent pipe types should have already been handled!");
     return QualType();
   }
+  case Type::TypeVariable:
```

Comment:
Could you add a comment to explain what is going on here?    Type variables are compatible only if they are identical.  That's covered by the earlier check that the canonical types are identical.  We've reached this point only if the type variables are not compatible, so we return QualType().